File: RetiledStart/libRetiledStart/AppsList.vb
' libRetiledStart - Utility library for RetiledStart written in VB.NET so that
'                   I don't have to write everything in C#.
' Copyright (C) 2021 Drew Naylor
' (Note that the copyright years include the years left out by the hyphen.)
' Windows Phone and all other related copyrights and trademarks are property
' of Microsoft Corporation. All rights reserved.
'
' This file is a part of the Retiled project.
' Neither Retiled nor Drew Naylor are associated with Microsoft
' and Microsoft does not endorse Retiled.
' Any other copyrights and trademarks belong to their
' respective people and companies/organizations.
'
'
'   Licensed under the Apache License, Version 2.0 (the "License");
'   you may not use this file except in compliance with the License.
'   You may obtain a copy of the License at
'
'     http://www.apache.org/licenses/LICENSE-2.0
'
'   Unless required by applicable law or agreed to in writing, software
'   distributed under the License is distributed on an "AS IS" BASIS,
'   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
'   See the License for the specific language governing permissions and
'   limitations under the License.



Imports libdotdesktop_standard
Public Class AppsList

    Public Shared Sub RunApp(ExecFilename As String)
        ' Runs what's passed to it.
        ' Try/Catch so that it doesn't crash if it can't find the program.
        Try
            ' Clean the Exec key and store it in a string so it's easy to access.
            ' It has to be in a List(Of String) so the command-line arguments
            ' are accessible.
            Dim CleanedExecKey As New List(Of String)
            CleanedExecKey = desktopEntryStuff.cleanExecKey(ExecFilename)
            'Debug.WriteLine(CleanedExecKey(0))
            'Debug.WriteLine(CleanedExecKey(1))

            If CleanedExecKey(1) IsNot Nothing Then
                ' Some apps such as GNOME Maps use command-line arguments
                ' in their Exec key, so we have to deal with those.
                ' This app doesn't want to use "gapplication launch".
                Process.Start(CleanedExecKey(0), CleanedExecKey(1))
            Else
                ' This particular app doesn't use command-line arguments.
                Process.Start(CleanedExecKey(0))
            End If
            'End If
        Catch ex As Exception
        End Try
    End Sub

    Public Shared Function GetDotDesktopFiles() As ObjectModel.ObservableCollection(Of DotDesktopEntryInAllAppsList)
        ' Gets all .desktop files in /usr/share/applications
        ' on Linux or my desktop on Windows.

        ' Define a collection of filenames to use.
        Dim DotDesktopFilesList As New List(Of DotDesktopEntryInAllAppsList)
        ' Define a path we'll set later.
        ' We're setting up a fallback, too.
        Dim DotDesktopFilesPath As String = "/usr/share/applications"

        If OperatingSystem.IsLinux = True Then
            DotDesktopFilesPath = "/usr/share/applications"

        ElseIf OperatingSystem.IsWindows = True Then
            'DotDesktopFilesPath = "C:\Users\Drew\Desktop"
            DotDesktopFilesPath = "C:\Users\drewn\Desktop"
        End If

        For Each DotDesktopFile As String In FileIO.FileSystem.GetFiles(DotDesktopFilesPath)
            ' Check if the file ends with .desktop.
            If DotDesktopFile.EndsWith(".desktop") Then

                If Not desktopEntryStuff.getInfo(DotDesktopFile, "NoDisplay") = "true" Then
                    ' Make sure this .desktop file is supposed to be shown.
                    ' Add the item after making sure the name exists.
                    If Not desktopEntryStuff.getInfo(DotDesktopFile.ToString, "Name") = String.Empty Then
                        DotDesktopFilesList.Add(New DotDesktopEntryInAllAppsList(DotDesktopFile.ToString,
                                                                                     desktopEntryStuff.getInfo(
                                                                                     DotDesktopFile.ToString, "Name"),
                                                                                     DotDesktopFile.ToString))
                    Else
                        DotDesktopFilesList.Add(New DotDesktopEntryInAllAppsList(DotDesktopFile.ToString,
                                                                                     DotDesktopFile.ToString,
                                                                                     DotDesktopFile.ToString))
                    End If
                End If

                End If
        Next

        ' This is where we actually sort the list.
        ' Stuff here ended up being really useful.
        ' Didn't know list items could have properties.
        ' Maybe one of my other programs that uses a List
        ' could benefit from this.
        ' https://stackoverflow.com/questions/11735902/sort-a-list-of-object-in-vb-net
        ' This answer in particular is what worked I think:
        ' https://stackoverflow.com/a/11736001
        DotDesktopFilesList = DotDesktopFilesList.OrderBy(Function(x) x.NameKeyValueProperty).ToList()

        ' Define a new ObservableCollection that we'll use to copy the file paths into.
        Dim NewDotDesktopFilesList As New ObjectModel.ObservableCollection(Of DotDesktopEntryInAllAppsList)

        ' Add all of the items that are file paths to the new ObservableCollection.
        For Each Item In DotDesktopFilesList
            NewDotDesktopFilesList.Add(New DotDesktopEntryInAllAppsList(Item.FileNameProperty, Item.NameKeyValueProperty,
                                                                        Item.ExecKeyValueProperty))
        Next

        ' Return the collection.
        Return NewDotDesktopFilesList

    End Function

    Public Shared Function GetDotDesktopNameKey(DotDesktopFile As String) As String
        ' Checks if the .desktop file actually has a "Name" key.
        If desktopEntryStuff.getInfo(DotDesktopFile, "Name") = String.Empty Then
            ' Return the filename.
            ' Preferably this wouldn't include the file path,
            ' but that would require using what dotbakker does
            ' where a file path is passed in and just the filename
            ' without the rest of the path is returned
            ' and it would take a little longer to get working.
            ' Had a bit of trouble figuring this out, but it turned out
            ' that it was returning String.Empty, rather than Nothing.
            'Debug.WriteLine(DotDesktopFile)
            Return DotDesktopFile
        Else
            ' Return what's in the "Name" key.
            'Debug.WriteLine(desktopEntryStuff.getInfo(DotDesktopFile, "Name"))
            Return desktopEntryStuff.getInfo(DotDesktopFile, "Name")
        End If
    End Function

End Class

Public Class DotDesktopEntryInAllAppsList
    ' Adding a new class so we can add them to a list
    ' so that sorting the list is easy.
    ' Details:
    ' https://docs.microsoft.com/en-us/dotnet/visual-basic/programming-guide/concepts/linq/how-to-create-a-list-of-items

    ' Properties:
    Public Property FileNameProperty As String
    Public Property NameKeyValueProperty As String
    Public Property ExecKeyValueProperty As String

    ' Required due to "Your custom class must be public and support a default (parameterless) public constructor."
    ' https://docs.microsoft.com/en-us/dotnet/desktop/wpf/advanced/xaml-and-custom-classes-for-wpf?view=netframeworkdesktop-4.8
    Public Sub New()

    End Sub

    Public Sub New(fileName As String,
                   nameKeyValue As String,
                   execKeyValue As String)
        ' Set the properties to be the parameters.
        FileNameProperty = fileName
        NameKeyValueProperty = nameKeyValue
        ' Add the exec key value, which is just the .desktop file.
        ExecKeyValueProperty = execKeyValue

    End Sub

End Class

File: RetiledStart/PyRetiledStart/libs/libRetiledStartPy/appslist.py
# libRetiledStartPy - Utility library for RetiledStart that allows me to
#                     keep the main code out of the UI code, like MVVM.
#                     This is a port of the VB.NET version to Python.
# Copyright (C) 2021 Drew Naylor
# (Note that the copyright years include the years left out by the hyphen.)
# Windows Phone and all other related copyrights and trademarks are property
# of Microsoft Corporation. All rights reserved.
#
# This file is a part of the Retiled project.
# Neither Retiled nor Drew Naylor are associated with Microsoft
# and Microsoft does not endorse Retiled.
# Any other copyrights and trademarks belong to their
# respective people and companies/organizations.
#
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

import subprocess
from ..libdotdesktop_py import desktopEntryStuff
# Stuff for getting the files from /usr/share/applications.
import os
from os.path import isfile, join

# Python allows relative imports as used above:
# https://stackoverflow.com/a/714647

# Trying to create a class and library similar to this one:
# https://codeigo.com/python/import-class-from-another-file-and-directory
# I'm also trying to make a thing that'll run apps based on the Python docs:
# https://docs.python.org/3/library/subprocess.html#popen-constructor
#class AppsList(object):
    # We have to do an __init__ thing:
	# https://careerkarma.com/blog/python-missing-required-positional-argument-self/
    #def init(self, ExecFilename):
        #self.ExecFilename = ExecFilename

# At least this works at all.
# TODO: Figure out how to use this with a class so that
# the code can be cleaner.
def RunApp(DotDesktopFilePath):
        # Get the ExecFilename split using shlex.split.
	args = desktopEntryStuff.cleanExecKey(DotDesktopFilePath)
	#print("Split: ")
	#print(args)
		# Now run the command.
		# TODO: Ensure the command is wrapped in quotes
		# if I already do that in my VB.NET library.
		# Actually, I'm not sure if I am, but that version
		# seems to do just fine when an Exec key lacks quotes
		# around the first part of the command, unlike this one
		# which says it can't find the file.
	proc = subprocess.Popen(args)

def GetAppName(DotDesktopFilePath):
	# Gets the app's name using the libdotdesktop_py library.
	return desktopEntryStuff.getInfo(DotDesktopFilePath, "Name", DotDesktopFilePath, "", True)
	
def getDotDesktopFiles():
	# Gets the list of .desktop files and creates a list of objects
	# of type DotDesktopEntryInAllAppsList that can be put into QML
	# after being ordered by name. The DotDesktopEntryInAllAppsList
	# type is a class, and that class has two properties:
	# FileNameProperty, which is the .desktop file path, and
	# NameKeyValueProperty, which stores the "Name" key value
	# to display the app's name in the All Apps list.
	# The FileNameProperty will be used to launch the apps by
	# passing it to desktopEntryStuff.getInfo.
	# Note: This description may be outdated as I'm currently
	# not using properties for the items, and just using a dictionary.
	
	
	# Get the list of files from /usr/share/applications:
	# https://stackoverflow.com/a/3207973
	# Only put apps in the list if they're supposed to be shown.
	# Using the example from this answer:
	# https://stackoverflow.com/a/51850082
	
	# Specify root path and slash.
	# This is different on Windows for debugging purposes.
	# Example code for sys.platform:
	# https://docs.python.org/3/library/sys.html#sys.platform
	if sys.platform.startswith("win32"):
		DotDesktopRootPath = "C:\\Users\\drewn\\Desktop"
		slash = "\\"
	else:
		DotDesktopRootPath = "/usr/share/applications"
		slash = "/"
		
	# Use the filesystem encode thing to get the folder.
	FSEncodedFolder = os.fsencode(DotDesktopRootPath)
	
	# Create empty list that will be written to later.
	DotDesktopFilesList = []
	
	# Loop through the files and add them to the list.
	for DotDesktopFile in os.listdir(FSEncodedFolder):
		DotDesktopFilename = os.fsdecode(DotDesktopFile)
		# Ensure only .desktop files are picked up.
		if DotDesktopFilename.endswith( (".desktop") ):
			# Make sure the .desktop file doesn't have NoDisplay = true.
			if not desktopEntryStuff.getInfo(DotDesktopRootPath + slash + DotDesktopFilename, "NoDisplay", "false", "", True) == "true":
				DotDesktopFilesList.append(DotDesktopFilename)

	# Put it back into a list because I don't know how to
	# use dictionaries with QML listview models yet.
	# Example here:
	# https://pythonexamples.org/python-dictionary-values-to-list/#4
	# This example may work better:
	# https://tutorial.eyehunts.com/python/python-add-to-dict-in-a-loop-adding-item-to-dictionary-within-loop-example/
	SortedDotDesktopFilesList = {}
	# This is going through the items in the DotDesktopFilesList and adding items to a dictionary by getting the
	# name from the .desktop file's "Name" key. If the "Name" key doesn't exist, it just uses the filename.
	for i in range(len(DotDesktopFilesList)):
		SortedDotDesktopFilesList[DotDesktopFilesList[i]] = desktopEntryStuff.getInfo(DotDesktopRootPath + slash + DotDesktopFilesList[i], "Name", DotDesktopFilesList[i], "", True)
	
	# Now we can sort the dictionary by values.
	# I think the filename sorting above will be
	# unnecessary with this, and it'll be removed.
	# Example from here:
	# https://www.30secondsofcode.org/python/s/sort-dict-by-value
	# This needs case-folding to ensure things are where they're
	# supposed to be:
	# https://stackoverflow.com/a/57923460
	# Note that the ".casefold()" has to be after the x[1] or it won't work.
	# Shorter example: "...key = lambda x: x[1].casefold()..."
	# If it's in the wrong spot, it might say "Error: 'tuple' object has no attribute 'casefold'"
	SortedDotDesktopFilesList = dict(sorted(SortedDotDesktopFilesList.items(), key = lambda x: x[1].casefold()))
	
	# Return a list for now, as using a dictionary may take longer to hook up:
	# https://www.tutorialspoint.com/How-to-get-a-list-of-all-the-keys-from-a-Python-dictionary
	# Maybe I can just re-use the old list instead of creating a new one.
	DotDesktopFilesList = [key for key in SortedDotDesktopFilesList]
	
	return DotDesktopFilesList